src/main/java/com/nvs/store/service/CartService.java

```diff
@@ -0,0 +1,202 @@
+import com.nvs.store.repository.*;
+import com.nvs.store.util.CartUtils;
+import jakarta.transaction.Transactional;
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+
+@Service
+@RequiredArgsConstructor
+public class CartService {
+    private final CartRepository cartRepository;
+    private final CartItemRepository cartItemRepository;
+    private final UserRepository userRepository;
+    private final ProductRepository productRepository;
+
+    public CartDTO findActualCartByEmail() {
+        String userEmail = getUserEmailFromContext();
+        User user = userRepository
+                .findByEmail(userEmail)
+                .orElseThrow(() -> new RuntimeException(
+                        String.format("User with email [%s] not found", userEmail))
+                );
+        Cart cart = cartRepository
+                .findByUserIdAndActualTrue(user.getId());
+        CartDTO cartDTO;
+        if (cart == null) {
+            Cart newCart = Cart.builder()
+                    .user(user)
+                    .actual(true)
+                    .cartItems(List.of())
+                    .build();
+            cartRepository.save(newCart);
+            cartDTO = CartUtils.convertToCartDto.apply(newCart);
+            return cartDTO;
+        }
+        cartDTO = CartUtils.convertToCartDto.apply(cart);
+        return cartDTO;
+    }
+
+    public List<CartDTO> getAllCarts() {
+        String userEmail = getUserEmailFromContext();
+        User user = userRepository
+                .findByEmail(userEmail)
+                .orElseThrow(() -> new RuntimeException(
+                        String.format("User with email [%s] not found", userEmail))
+                );
+        return cartRepository
+                .findByUserId(user.getId())
+                .stream()
+                .map(CartUtils.convertToCartDto)
+                .toList();
+    }
+
+    public CartDTO addProductToCart(CartItemRequest cartItemRequest) {
+        String userEmail = getUserEmailFromContext();
+        User user = userRepository.findByEmail(userEmail)
+                .orElseThrow(() -> new RuntimeException(
+                        String.format("User with username %s not found", userEmail))
+                );
+
+        Product product = productRepository
+                .findById(cartItemRequest
+                        .getProductId())
+                .orElseThrow(() ->
+                        new RuntimeException(
+                                String.format("Product with id [%d] does not exist", cartItemRequest.getProductId()))
+                );
+        Cart cart = cartRepository
+                .findByUserIdAndActualTrue(user.getId());
+
+        List<CartItem> cartItems = cartItemRepository.findCartItemByCart(cart);
```

Comment:
When the cart is loaded from database, the cartItems are populated. You can use the CartItem from cart instead of loading one more time the cart items by separate select, it looks like a redundant operation. The cart items are loaded because the fetch loading is set to eager.

Reply:
Done ✅